sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaMetrics.java

```diff
@@ -0,0 +1,126 @@
+      return singleton;
+    }
+  }
+
+  /**
+   * Metrics of a batch of RPCs. Member variables are thread safe; however, this class does not have
+   * atomicity across member variables.
+   *
+   * <p>Expected usage: A number of threads record metrics in an instance of this class with the
+   * member methods. Afterwards, a single thread should call {@code updateStreamingInsertsMetrics}
+   * which will export all counters metrics and RPC latency distribution metrics to the underlying
+   * {@code perWorkerMetrics} container. Afterwards, metrics should not be written/read from this
+   * object.
+   */
+  @AutoValue
+  abstract class KafkaMetricsImpl implements KafkaMetrics {
+
+    static HashMap<String, Histogram> latencyHistograms = new HashMap<String, Histogram>();
+
+    abstract HashMap<String, ConcurrentLinkedQueue<Duration>> perTopicRpcLatencies();
+
+    abstract AtomicBoolean isWritable();
+
+    public static KafkaMetricsImpl create() {
+      return new AutoValue_KafkaMetrics_KafkaMetricsImpl(
+          new HashMap<String, ConcurrentLinkedQueue<Duration>>(), new AtomicBoolean(true));
+    }
+
+    /** Record the rpc status and latency of a successful Kafka poll RPC call. */
+    @Override
+    public void updateSuccessfulRpcMetrics(String topic, Duration elapsedTime) {
+      if (isWritable().get()) {
+        ConcurrentLinkedQueue<Duration> latencies = perTopicRpcLatencies().get(topic);
+        if (latencies == null) {
+          latencies = new ConcurrentLinkedQueue<Duration>();
+          latencies.add(elapsedTime);
+          perTopicRpcLatencies().put(topic, latencies);
+        } else {
+          latencies.add(elapsedTime);
+        }
+      }
+    }
+
+    /** Record rpc latency histogram metrics for all recorded topics. */
+    private void recordRpcLatencyMetrics() {
+      for (Map.Entry<String, ConcurrentLinkedQueue<Duration>> topicLatencies :
+          perTopicRpcLatencies().entrySet()) {
+        Histogram topicHistogram;
+        if (latencyHistograms.containsKey(topicLatencies.getKey())) {
+          topicHistogram = latencyHistograms.get(topicLatencies.getKey());
+        } else {
+          topicHistogram =
+              KafkaSinkMetrics.createRPCLatencyHistogram(
+                  KafkaSinkMetrics.RpcMethod.POLL, topicLatencies.getKey());
+          latencyHistograms.put(topicLatencies.getKey(), topicHistogram);
+        }
+        // update all the latencies
+        for (Duration d : topicLatencies.getValue()) {
+          Preconditions.checkArgumentNotNull(topicHistogram);
+          topicHistogram.update(d.toMillis());
+        }
+      }
+    }
+
+    /**
+     * Export all metrics recorded in this instance to the underlying {@code perWorkerMetrics}
+     * containers. This function will only report metrics once per instance. Subsequent calls to
+     * this function will no-op.
+     */
+    @Override
+    public void updateKafkaMetrics() {
+      if (!isWritable().compareAndSet(true, false)) {
```

Comment:
Should this "fail" silently? Id assume that if this is called twice, something has gone wrong

Reply:
Added a warning log, but the way its called in the Unbounded Reader, we can't call this method twice.

Reply:
sg